rust/src/dataset.rs

```diff
@@ -152,8 +154,28 @@ impl Dataset {
             ));
         }
 
-        let mut fragment_id = 0;
-        let mut fragments: Vec<Fragment> = vec![];
+        if matches!(params.mode, WriteMode::Append) {
+            if let Some(m) = latest_manifest.as_ref() {
+                if schema != m.schema {
+                    return Err(Error::IO(format!(
+                        "Append with different schema: original={} new={}",
+                        m.schema, schema
+                    )));
+                }
+            }
+        }
+
+        let mut fragment_id = latest_manifest.as_ref().map_or(0, |m| {
```

Comment:
would it make sense to put max fragment in manifest?

Reply:
This is interesting about how to enforce it. i.e., allow a background thread to start compact but allow writing, or just hard stop users to append new fragments. 

This could be a question we need to answer when the scale is up.